vm/devices/storage/ide/src/drive/atapi_drive.rs

```diff
@@ -455,8 +451,11 @@ impl AtapiDrive {
     }
 
     fn complete_data_port_read(&mut self) {
-        self.remove_status(Status::BSY | Status::ERR | Status::DRQ);
-        self.insert_status(Status::DRDY | Status::DSC);
+        self.state.regs.status.set_bsy(false);
```

Comment:
I'm a little concerned about atomicity here. I'll think further, but is it conceivable that there's now a potential for torn state that races with these updates?

Reply:
Relatively sure the &mut self means that we have exclusive access here.

Reply:
Fair, and apologies - it takes me some time to get through this. The guest can read these bits at any point. Are you asserting that the emulated mmio read cannot execute concurrently? (I think that's what you're saying, in which case my concerns are allayed).

Reply:
Yes, that's correct. That's guaranteed both by the &mut self here, and that all devices are behind a Mutex at a higher level (to enable the &mut self here).